### PR TITLE
Use HistoryReader for navigation replay

### PR DESCRIPTION
**Pull request by kmadsen**
### Description

Resolves https://github.com/mapbox/mapbox-navigation-android/issues/4485

Adding the ability to replay pbf.gz files. Deleting the old path to replay only json files. This will continue to support the old json format, but you have to pass the file path to the `MapboxHistoryReader`.

I have not found a way to read the `*.pbf.gz` formatted files in the assets or raw resource directories. I've had to copy the files into my device storage, and then read the file directly from there. 

### Changelog

```
<changelog>
Delete `ReplayHistoryDTO` in favor of `MapboxHistoryReader`.
Add eventTimestamp to all `HistoryEvent` classes.
Delete `CustomEventMapper` in favor of `ReplayHistoryEventMapper`
Create new `ReplayHistoryMapper` which is extendable and customizable
</changelog>
```

### Screenshots or Gifs

``` kotlin
val mapper = ReplayHistoryMapper.Builder()
    .statusMapper(null)
    .addPushEventMapper(fooEventMapper)
    .addPushEventMapper(initialRouteMapper)
    .build()

val replayEvents = MapboxHistoryReader(outputFile.absolutePath).asSequence()
    .mapNotNull { mapper.mapToReplayEvent(it) }
    .toList()

mapboxReplayer.clearEvents()
mapboxReplayer.pushEvents(replayEvents)
mapboxReplayer.play()
```

https://user-images.githubusercontent.com/3021882/126387047-8be9c46e-8a64-4fc4-a487-9e1283f02fb1.mp4

### CODE REVIEW

**Comment by kmadsen** [comment]
Robo tests seem to be failing because of this https://github.com/mapbox/mapbox-navigation-native/issues/3583

File: examples/src/main/java/com/mapbox/navigation/examples/core/replay/HistoryFilesDirectory.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        replayDirectory(context).walk().filterNot { it.isDirectory }.toList()
+
+    /**
+     * Returns a file in the [replayDirectory] where a history file can be written.
+     */
+    fun outputFile(context: Context, path: String) =
+        File(replayDirectory(context), path)
+
+    private companion object {
+        const val DIRECTORY_NAME = "replay"
```
[the remark this answers is not in the document]
Comment (kmadsen) [reply examples/src/main/java/com/mapbox/navigation/examples/core/replay/HistoryFilesDirectory.kt:32]: defined the companion object private, so this field is also private

File: examples/src/main/java/com/mapbox/navigation/examples/core/replay/HistoryFilesDirectory.kt [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package com.mapbox.navigation.examples.core.replay
+
+import android.content.Context
+import java.io.File
+
+/**
+ * Helper class to keep the a consistent directory.
```
[the remark this answers is not in the document]
Comment (kmadsen) [reply examples/src/main/java/com/mapbox/navigation/examples/core/replay/HistoryFilesDirectory.kt]: thanks!

File: examples/src/main/java/com/mapbox/navigation/examples/core/replay/HistoryFilesDirectory.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Guardiola31337) [reply examples/src/main/java/com/mapbox/navigation/examples/core/replay/HistoryFilesDirectory.kt:31]: Not really. Look at the decompiled code 👀 

_Before_

```java
   /** @deprecated */
   @Deprecated
   @NotNull
   public static final String DIRECTORY_NAME = "replay";
```

_After_

```java
   /** @deprecated */
   @Deprecated
   private static final String DIRECTORY_NAME = "replay";
```

**Review by Guardiola31337 — COMMENTED** [review]
Left some minor comments @kmadsen

File: examples/src/main/java/com/mapbox/navigation/examples/core/replay/HistoryFilesDirectory.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (kmadsen) [reply examples/src/main/java/com/mapbox/navigation/examples/core/replay/HistoryFilesDirectory.kt:31]: hmm well that's fair. wonder if that's intentional

**Review by Guardiola31337 — APPROVED** [review]
Thanks for addressing the feedback 🚀

**Pull request merged**